File: 0001-0500/0039/0039.combination-sum.js
/*
 * @lc app=leetcode id=39 lang=javascript
 *
 * [39] Combination Sum
 */

// @lc code=start
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum = function (candidates, target) {
    let result = [];
    let current = [];
    candidates.sort((a, b) => a - b);
    helper(candidates, target, 0, current, result);
    return result;
};

var helper = function (candidates, target, start, current, result) {
    if (target === 0) {
        result.push(current.slice());
        return;
    }
    for (let i = start; i < candidates.length; i++) {
        if (candidates[i] > target) {
            break;
        }
        current.push(candidates[i]);
        helper(candidates, target - candidates[i], i, current, result);
        current.pop();
    }
};
// @lc code=end

File: 0001-0500/0039/0039.combination-sum.cs
/*
 * @lc app=leetcode id=39 lang=csharp
 *
 * [39] Combination Sum
 */

// @lc code=start
public class Solution {
    public IList<IList<int>> CombinationSum(int[] candidates, int target) {
        IList<IList<int>> result = new List<IList<int>>();
        List<int> combination = new List<int>();
        Array.Sort(candidates);
        Backtrack(candidates, target, result, combination, 0);
        return result;
    }
    
    private void Backtrack(int[] candidates, int target, IList<IList<int>> result, List<int> combination, int start) {
        if (target == 0) {
            result.Add(new List<int>(combination));
            return;
        }
        
        for (int i = start; i < candidates.Length; i++) {
            if (candidates[i] > target) break;
            combination.Add(candidates[i]);
            Backtrack(candidates, target - candidates[i], result, combination, i);
            combination.RemoveAt(combination.Count - 1);
        }
    }
}
// @lc code=end

